src/Microsoft.DotNet.Interactive.Tests/LanguageServices/CompletionTests.cs

```diff
@@ -0,0 +1,117 @@
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.DotNet.Interactive.Commands;
+using Microsoft.DotNet.Interactive.Events;
+using Microsoft.DotNet.Interactive.Tests.Utility;
+using Xunit;
+using Xunit.Abstractions;
+
+#pragma warning disable 8509
+namespace Microsoft.DotNet.Interactive.Tests.LanguageServices
+{
+    public class CompletionTests : LanguageKernelTestBase
+    {
+        public CompletionTests(ITestOutputHelper output) : base(output)
+        {
+        }
+
+        [Theory]
+        [InlineData(Language.FSharp)]
+        [InlineData(Language.CSharp)]
+        public async Task Completions_are_available_for_symbols_declared_in_the_previous_submission(Language language)
+        {
+            var variableName = "aaaaaaa";
+
+            var declarationSubmission = language switch
+            {
+                Language.CSharp => $"var {variableName} = 123;",
+                Language.FSharp => $"let {variableName} = 123"
+            };
+
+            var kernel = CreateKernel(language);
+
+            await kernel.SubmitCodeAsync(declarationSubmission);
+
+            await kernel.SendAsync(new RequestCompletion("aaa", new LinePosition(0, 3)));
+
+            KernelEvents
+                .Should()
+                .ContainSingle<CompletionRequestCompleted>()
+                .Which
+                .CompletionList
+                .Should()
+                .Contain(item => item.DisplayText == variableName);
+        }
+
+        [Theory]
+        [InlineData(Language.FSharp)]
+        [InlineData(Language.CSharp)]
+        public async Task Completions_are_available_for_symbols_declared_in_a_submission_before_the_previous_submission(Language language)
+        {
+            var variableName = "aaaaaaa";
+
+            var submissions = language switch
+            {
+                Language.CSharp => new[]
+                {
+                    $"var {variableName} = 123;",
+                    $"var bbbbb = 456;"
+                },
+                Language.FSharp => new[]
+                {
+                    $"let {variableName} = 123",
+                    $"let bbbbb = 456"
+                }
+            };
+
+            var kernel = CreateKernel(language);
+
+            foreach (var submission in submissions)
+            {
+                await kernel.SubmitCodeAsync(submission);
+            }
+
+            await kernel.SendAsync(new RequestCompletion("aaa", new LinePosition(0, 2)));
+
+            KernelEvents
+                .Should()
+                .ContainSingle<CompletionRequestCompleted>()
+                .Which
+                .CompletionList
+                .Should()
+                .Contain(item => item.DisplayText == variableName);
+        }
+
+        [Theory]
+        [InlineData(Language.FSharp)]
+        [InlineData(Language.CSharp)]
+        public async Task Completions_are_available_for_symbols_declared_in_a_previous_submission_ending_in_a_trailing_expression(Language language)
```

Comment:
What about symbols being available 2+ submissions back, each with trailing expressions?  I want to make sure we don't miss that scenario, because it's very useful to have each cell end with an expression to see the state as you're building your notebook.

Reply:
I'll add more scenarios to these tests once the current ones are passing. Also on the list, submissions containing magic commands.